

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Quick Start Example &mdash; iMARGI Pipeline 1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Step-by-step Illustration" href="step_by_step_illustration.html" />
    <link rel="prev" title="Tools and Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> iMARGI Pipeline
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Overview</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Tools and Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start Example</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pull-the-imargi-docker-image">Pull the iMARGI Docker Image</a></li>
<li class="toctree-l2"><a class="reference internal" href="#example-data-preparation">Example Data Preparation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#set-working-directory">Set Working Directory</a></li>
<li class="toctree-l3"><a class="reference internal" href="#prepare-fastq-files-of-paired-end-sequencing">Prepare FASTQ Files of Paired-end Sequencing</a><ul>
<li class="toctree-l4"><a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/sample_R1.fastq.gz">Click Here to Download: R1 reads fastq.gz</a></li>
<li class="toctree-l4"><a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/sample_R2.fastq.gz">Click Here to Download: R2 reads fastq.gz</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#download-reference-genome">Download Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-bwa-index-files">Download BWA Index Files</a><ul>
<li class="toctree-l4"><a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/bwa_index.tar.gz">Click Here to Download: bwa index files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#checklist-of-preparation">Checklist of Preparation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#perform-imargi-pipeline-in-one-command">Perform iMARGI Pipeline in One Command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-of-imargi-pipeline">Output of iMARGI Pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="step_by_step_illustration.html">Step-by-step Illustration</a></li>
<li class="toctree-l1"><a class="reference internal" href="further_analysis.html">Further Analysis and Visualization Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="technical_note.html">Technical Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq_win_mac.html">Guides for Issues on Windows and macOS System</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_api.html">Command-line API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">iMARGI Pipeline</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Quick Start Example</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="quick-start-example">
<h1>Quick Start Example<a class="headerlink" href="#quick-start-example" title="Permalink to this headline">¶</a></h1>
<p>This quick start example shows how to apply the iMARGI Pipeline using iMARGI-Docker. The all-in-one wrapper script
<code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> is used here to automate the whole pipeline in one command line.
Please go to the <a class="reference internal" href="step_by_step_illustration.html"><span class="doc">Step-by-step Illustration</span></a> section to learn the details of
the pipeline.</p>
<ul class="simple">
<li><a class="reference external" href="#quick-start-example">Quick Start Example</a><ul>
<li><a class="reference external" href="#pull-the-imargi-docker-image">Pull the iMARGI Docker Image</a></li>
<li><a class="reference external" href="#example-data-preparation">Example Data Preparation</a><ul>
<li><a class="reference external" href="#set-working-directory">Set Working Directory</a></li>
<li><a class="reference external" href="#prepare-fastq-files-of-paired-end-sequencing">Prepare FASTQ Files of Paired-end Sequencing</a></li>
<li><a class="reference external" href="#download-reference-genome">Download Reference Genome</a></li>
<li><a class="reference external" href="#download-bwa-index-files">Download BWA Index Files</a></li>
<li><a class="reference external" href="#checklist-of-preparation">Checklist of Preparation</a></li>
</ul>
</li>
<li><a class="reference external" href="#perform-imargi-pipeline-in-one-command">Perform iMARGI Pipeline in One Command</a></li>
<li><a class="reference external" href="#output-of-imargi-pipeline">Output of iMARGI Pipeline</a></li>
</ul>
</li>
</ul>
<div class="section" id="pull-the-imargi-docker-image">
<h2>Pull the iMARGI Docker Image<a class="headerlink" href="#pull-the-imargi-docker-image" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>docker pull zhonglab/imargi
</pre></div>
</div>
</div>
<div class="section" id="example-data-preparation">
<h2>Example Data Preparation<a class="headerlink" href="#example-data-preparation" title="Permalink to this headline">¶</a></h2>
<div class="section" id="set-working-directory">
<h3>Set Working Directory<a class="headerlink" href="#set-working-directory" title="Permalink to this headline">¶</a></h3>
<p>We need to mount a directory on host machine to Docker container as working directory. Here we set the
<code class="docutils literal notranslate"><span class="pre">~/imargi_example</span></code> directory as the mounted working directory, which means we need to use <code class="docutils literal notranslate"><span class="pre">-v</span> <span class="pre">~/imargi_example:/imargi</span></code>
option when we run iMARGI-Docker. We created some sub-directories for organizing data and results.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/imargi_example
mkdir ref data output
</pre></div>
</div>
<p>Then the structure of working directory is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~/imargi_example
    ├── ref
    │   └──
    ├── data
    │   └──
    └── output
        └──
</pre></div>
</div>
</div>
<div class="section" id="prepare-fastq-files-of-paired-end-sequencing">
<h3>Prepare FASTQ Files of Paired-end Sequencing<a class="headerlink" href="#prepare-fastq-files-of-paired-end-sequencing" title="Permalink to this headline">¶</a></h3>
<p>Here we take a small demo dataset as the example. The demo dataset can be processed in about 15 mins.
Here are the download links:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/sample_R1.fastq.gz">Click Here to Download: R1 reads fastq.gz</a></li>
<li class="toctree-l1"><a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/sample_R2.fastq.gz">Click Here to Download: R2 reads fastq.gz</a></li>
</ul>
</div>
<p>The demo dataset is randomly sampled from a real HiSeq iMARGI data in HEK cells generated in our previous published
paper <a id="a1"><a class="reference external" href="#f1">[1]</a></a>. If you are interested in the real dataset, you can get it in NCBI SRA database with
accession number <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra?term=SRX5026004">SRR8206679</a>. It’s a very large dataset
(more than 40 GB) and processing it will take a lot of time (more than 10 hours).</p>
</div>
<div class="section" id="download-reference-genome">
<h3>Download Reference Genome<a class="headerlink" href="#download-reference-genome" title="Permalink to this headline">¶</a></h3>
<p>The minimum required reference file is only a reference genome sequence FASTA file.</p>
<p>We use the same reference genome used by
<a class="reference external" href="https://www.4dnucleome.org/">4D Nucleome</a> and
<a class="reference external" href="https://www.encodeproject.org/data-standards/reference-sequences/">ENCODE project</a>. The FASTA file of the reference
genome can be downloaded use the link:</p>
<p><a class="reference external" href="https://www.encodeproject.org/files/GRCh38_no_alt_analysis_set_GCA_000001405.15/&#64;&#64;download/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz">Click Here to Download: GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz</a>.</p>
<p>It needs to be decompressed using <code class="docutils literal notranslate"><span class="pre">gunzip</span> <span class="pre">-d</span></code> command. The command lines for preparing the reference genome files:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/imargi_example

wget https://www.encodeproject.org/files/GRCh38_no_alt_analysis_set_GCA_000001405.15/@@download/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz
gunzip -d GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz
mv GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta ./ref
</pre></div>
</div>
</div>
<div class="section" id="download-bwa-index-files">
<h3>Download BWA Index Files<a class="headerlink" href="#download-bwa-index-files" title="Permalink to this headline">¶</a></h3>
<p>Although <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> can automatically build bwa index files, it will cost a lot of time (more than 1 hour
for human genome). Hence, we suggest to download our pre-built BWA index files for the human hg38 reference
genome and supply it to <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> with <code class="docutils literal notranslate"><span class="pre">-i</span></code> option. Please download the following gzip compressed <code class="docutils literal notranslate"><span class="pre">bwa_index</span></code>
folder and decompress it (<code class="docutils literal notranslate"><span class="pre">tar</span> <span class="pre">zxvf</span></code>) in your machine.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/bwa_index.tar.gz">Click Here to Download: bwa index files</a></li>
</ul>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/imargi_example/ref
wget  https://sysbio.ucsd.edu/imargi_pipeline/bwa_index.tar.gz
tar zxvf bwa_index.tar.gz
</pre></div>
</div>
</div>
<div class="section" id="checklist-of-preparation">
<h3>Checklist of Preparation<a class="headerlink" href="#checklist-of-preparation" title="Permalink to this headline">¶</a></h3>
<p>After preparation, you will get all the required data and reference files in the working directory. It’s the minimum
requirements, which are sequencing reads data in FASTQ format and reference genome sequence in FASTA format.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~/imargi_example
    ├── data
    │   ├── sample_R1.fastq.gz
    │   └── sample_R2.fastq.gz
    ├── output
    └── ref
        ├── GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta
        └── bwa_index
            ├── bwa_index_hg38.amb
            ├── bwa_index_hg38.ann
            ├── bwa_index_hg38.bwt
            ├── bwa_index_hg38.pac
            └── bwa_index_hg38.sa
</pre></div>
</div>
</div>
</div>
<div class="section" id="perform-imargi-pipeline-in-one-command">
<h2>Perform iMARGI Pipeline in One Command<a class="headerlink" href="#perform-imargi-pipeline-in-one-command" title="Permalink to this headline">¶</a></h2>
<p>The easiest way of is using the all-in-one wrapper script <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> with default parameters.
We used <code class="docutils literal notranslate"><span class="pre">-N</span> <span class="pre">test_sample</span></code> argument to set base identifier for all the result files.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/imargi_example
docker run --rm -u <span class="m">1043</span> -v ~/imargi_example:/imargi zhonglab/imargi <span class="se">\</span>
    imargi_wrapper.sh <span class="se">\</span>
    -r hg38 <span class="se">\</span>
    -N test_sample <span class="se">\</span>
    -t <span class="m">4</span> <span class="se">\</span>
    -g ./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta <span class="se">\</span>
    -i ./ref/bwa_index/bwa_index_hg38 <span class="se">\</span>
    -1 ./data/sample_R1.fastq.gz <span class="se">\</span>
    -2 ./data/sample_R2.fastq.gz <span class="se">\</span>
    -o ./output
</pre></div>
</div>
<p>It will take about 10 minutes to complete the whole processing pipeline to the demo dataset (with <code class="docutils literal notranslate"><span class="pre">-i</span></code> bwa index argument).</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Step</th>
<th>Time</th>
<th>Speed up suggestion</th>
</tr>
</thead>
<tbody>
<tr>
<td>Generating chromosome size file</td>
<td>10 sec</td>
<td>It's fast, but you can also supply with <code>-c</code> once you've generated it before.</td>
</tr>
<tr>
<td>Generating bwa index (skipped)</td>
<td>75 min</td>
<td>Supply with <code>-i</code> if you've pre-built index files.</td>
</tr>
<tr>
<td>Generating restriction fragment file</td>
<td>4 min</td>
<td>Supply with <code>-R</code> when you've already created it before.</td>
</tr>
<tr>
<td>cleaning</td>
<td>10 sec</td>
<td>It's fast and not parallelization.</td>
</tr>
<tr>
<td>bwa mapping</td>
<td>2 min</td>
<td>More CPU cores with <code>-t</code>.</td>
</tr>
<tr>
<td>interaction pair parsing</td>
<td>1 min</td>
<td>More CPU cores with <code>-t</code>.</td>
</tr>
</tbody>
</table><blockquote>
<div><p><em>Tips:</em></p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">--rm</span></code>: By default a container’s file system persists even after the container exits. Hence, the container file
systems can really pile up. <code class="docutils literal notranslate"><span class="pre">--rm</span></code> option will automatically clean up the container after the container exits.</li>
<li><code class="docutils literal notranslate"><span class="pre">-u</span> <span class="pre">1043</span></code>: Run docker with your own UID of your Linux system (use <code class="docutils literal notranslate"><span class="pre">id</span></code> command to check your own UID and replace
<code class="docutils literal notranslate"><span class="pre">1043</span></code> with it) to avoid file/dir permission problem.</li>
<li><code class="docutils literal notranslate"><span class="pre">-v</span> <span class="pre">~/imargi_example:/imargi</span></code>: It mounts the <code class="docutils literal notranslate"><span class="pre">~/imargi_example</span></code> directory in your host machine to workspace of the
running docker container. The path must be a full path. The example was ran on a Linux computer. If you ran it on a
Windows  computer, the path is a little different. For example, Windows path <code class="docutils literal notranslate"><span class="pre">D:\test\imargi_example</span></code> needs to be
rewritten as <code class="docutils literal notranslate"><span class="pre">/d/test/imargi_example</span></code>, so the <code class="docutils literal notranslate"><span class="pre">-v</span></code> argument needs to be <code class="docutils literal notranslate"><span class="pre">-v</span> <span class="pre">/d/test/imargi_example:/imargi</span></code>. When you
executed it on Windows, a window might pop up to verify that you want to share the folder.</li>
<li>The command line is long, so <code class="docutils literal notranslate"><span class="pre">\</span></code> was used for splitting it into multiple lines in the example. It’s a Linux or MacOS
style. However, in Windows, you might need to replace <code class="docutils literal notranslate"><span class="pre">\</span></code> with <code class="docutils literal notranslate"><span class="pre">^</span></code>.</li>
<li><code class="docutils literal notranslate"><span class="pre">-i</span></code>: Building bwa index will cost a lot time, so we used the pre-built index files with <code class="docutils literal notranslate"><span class="pre">-i</span></code> argument. There
are also some other arguments can be used for pre-generated files, such as <code class="docutils literal notranslate"><span class="pre">-R</span></code> for restriction fragment BED file and
<code class="docutils literal notranslate"><span class="pre">-c</span></code> for chromsize file.See more details in the
<a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/commandline_api.html#imargi-wrapper-sh">documentation of command line API section</a></li>
<li><code class="docutils literal notranslate"><span class="pre">-i</span></code>: If you don’t supply bwa index files, the <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> will generated     it automatically. It works
perfectly on Linux system. However, it doesn’t work on Windows and MacOS because <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">index</span></code> use <code class="docutils literal notranslate"><span class="pre">fsync</span></code> when build
large genome index, which cannot handle different driver formats (<code class="docutils literal notranslate"><span class="pre">-v</span></code> mount Windows/MacOS driver to Linux container).
So it’s better to build it in advance. In fact, there’s a solution to the problem if you are familiar with Docker
volume. Please read the
<a class="reference external" href="https://sysbio.ucsd.edu/imargi_pipeline/technical_note.html#solve-bwa-index-failure-problem">technical note of iMARGI pipeline documentation</a> for
detail.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="output-of-imargi-pipeline">
<h2>Output of iMARGI Pipeline<a class="headerlink" href="#output-of-imargi-pipeline" title="Permalink to this headline">¶</a></h2>
<p>Once the iMARGI Pipeline has completed, all the result files are in the output directory. The tree structure of the
whole working directory is shown below. The final RNA-DNA interaction map is in the <code class="docutils literal notranslate"><span class="pre">final_test_sample.pairs.gz</span></code> file,
which is a compressed .pairs format file and can be used for further analysis. All the intermediate result files are
also kept in several sub-directories. As we used the minimum input requirements, so the pipeline automatically
generated several new reference files in the same directory of reference genome, including chromosome sizes, bwa index
and AluI digestion fragments. When processing new dataset, you can reuse these new generated reference files with
corresponding arguments instead of only using <code class="docutils literal notranslate"><span class="pre">-g</span></code> argument, which will save you some time and disk space. For more
detail, please check the <a class="reference internal" href="step_by_step_illustration.html"><span class="doc">Step-by-step Illustration</span></a> and
<a class="reference internal" href="commandline_api.html"><span class="doc">Command-line API</span></a> sections.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~/imargi_example/
    ├── data
    │   ├── sample_R1.fastq.gz
    │   └── sample_R2.fastq.gz
    ├── ref
    │   ├── GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta
    │   ├── GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.fai
    │   ├── chromsize.hg38.txt
    │   ├── AluI_frags.bed.gz
    │   ├── AluI_frags.bed.gz.tbi
    │   └── bwa_index
    │       ├── bwa_index_hg38.amb
    │       ├── bwa_index_hg38.ann
    │       ├── bwa_index_hg38.bwt
    │       ├── bwa_index_hg38.pac
    │       └── bwa_index_hg38.sa
    └── output
        ├── bwa_output
        │   ├── bwa_log_test_sample.txt
        │   └── test_sample.bam
        ├── clean_fastq
        │   ├── clean_test_sample_R1.fastq.gz
        │   └── clean_test_sample_R2.fastq.gz
        ├── parse_temp
        │   ├── dedup_test_sample.pairs.gz
        │   ├── drop_test_sample.pairs.gz
        │   ├── duplication_test_sample.pairs.gz
        │   ├── sorted_all_test_sample.pairs.gz
        │   ├── stats_dedup_test_sample.txt
        │   ├── stats_final_test_sample.txt
        │   └── unmapped_test_sample.pairs.gz
        ├── final_test_sample.pairs.gz
        └── pipelineStats_test_sample.log
</pre></div>
</div>
<p><strong>Reference:</strong></p>
<p><small><a class="reference external" href="#a1">[1]</a> <span id="f1"></span> Yan, Z. et al. Genome-wide co-localization of RNA-DNA interactions and fusion RNA pairs. PNAS February 19, 2019, 116 (8) 3328-3337. https://doi.org/10.1073/pnas.1819788116</small></p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="step_by_step_illustration.html" class="btn btn-neutral float-right" title="Step-by-step Illustration" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Tools and Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Zhong Lab, UCSD. iMARGI-Docker code licensed under the BSD 2 License. Documentation licensed under CC BY 3.0

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>