

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Step-by-step Illustration &mdash; iMARGI Pipeline 1.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Further Analysis and Visualization Guides" href="further_analysis.html" />
    <link rel="prev" title="Quick Start Example" href="quick_example.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> iMARGI Pipeline
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Overview</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Tools and Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_example.html">Quick Start Example</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Step-by-step Illustration</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#the-imargi-wrapper-sh-running-command">The <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> Running Command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-required-reference-files">Generating Required Reference Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#generating-chromosome-sizes-file">Generating Chromosome Sizes File</a></li>
<li class="toctree-l3"><a class="reference internal" href="#building-bwa-index">Building BWA Index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generating-restriction-fragment-file">Generating Restriction Fragment File</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cleaning-clean-read-pairs">Cleaning: clean read pairs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mapping-map-reads-to-genome">Mapping: map reads to genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parsing-parse-mapped-read-pairs-to-valid-rna-dna-interactions">Parsing: parse mapped read pairs to valid RNA-DNA interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">Output Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#all-the-output-file-list">All The Output File List</a></li>
<li class="toctree-l3"><a class="reference internal" href="#the-imargi-pipeline-output-pairs-format">The iMARGI Pipeline Output <code class="docutils literal notranslate"><span class="pre">.pairs</span></code> Format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#the-pipeline-summary-log-file">The pipeline summary log file</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="further_analysis.html">Further Analysis and Visualization Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="technical_note.html">Technical Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq_win_mac.html">Guides for Issues on Windows and macOS System</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_api.html">Command-line API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">iMARGI Pipeline</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Step-by-step Illustration</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="step-by-step-illustration">
<h1>Step-by-step Illustration<a class="headerlink" href="#step-by-step-illustration" title="Permalink to this headline">¶</a></h1>
<p>In the <a class="reference internal" href="quick_example.html"><span class="doc">Quick Start Example</span></a> section, we have shown the simplest all-in-one command line of
performing the pipeline with a wrapper script <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code>. Here we go inside of the quick example to illustrate
the details of iMARGI Pipeline. If you want to customize your pipeline, you can run each step separately with
corresponding tools.</p>
<p><img alt="_images/iMARGI_docker.png" src="_images/iMARGI_docker.png" />
<i><center>Schematic overview of the iMARGI data analysis pipeline </i></center></p>
<ul class="simple">
<li><a class="reference external" href="#step-by-step-illustration">Step-by-step Illustration</a><ul>
<li><a class="reference external" href="#the-imargi_wrappersh-running-command">The <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> Running Command</a></li>
<li><a class="reference external" href="#generating-required-reference-files">Generating Required Reference Files</a><ul>
<li><a class="reference external" href="#generating-chromosome-sizes-file">Generating Chromosome Sizes File</a></li>
<li><a class="reference external" href="#building-bwa-index">Building BWA Index</a></li>
<li><a class="reference external" href="#generating-restriction-fragment-file">Generating Restriction Fragment File</a></li>
</ul>
</li>
<li><a class="reference external" href="#cleaning-clean-read-pairs">Cleaning: clean read pairs</a></li>
<li><a class="reference external" href="#mapping-map-reads-to-genome">Mapping: map reads to genome</a></li>
<li><a class="reference external" href="#parsing-parse-mapped-read-pairs-to-valid-rna-dna-interactions">Parsing: parse mapped read pairs to valid RNA-DNA interactions</a></li>
<li><a class="reference external" href="#output-files">Output Files</a><ul>
<li><a class="reference external" href="#all-the-output-file-list">All The Output File List</a></li>
<li><a class="reference external" href="#the-imargi-pipeline-output-pairs-format">The iMARGI Pipeline Output <code class="docutils literal notranslate"><span class="pre">.pairs</span></code> Format</a></li>
<li><a class="reference external" href="#the-pipeline-summary-log-file">The pipeline summary log file</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="section" id="the-imargi-wrapper-sh-running-command">
<h2>The <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> Running Command<a class="headerlink" href="#the-imargi-wrapper-sh-running-command" title="Permalink to this headline">¶</a></h2>
<p>In the <a class="reference internal" href="quick_example.html"><span class="doc">quick example</span></a> section, we use the following command of <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code>, whose input
are only sequencing read pairs FASTQ files and reference genome sequence FASTA file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>docker run --rm -t -u <span class="m">1043</span> -v ~/imargi_example:/imargi zhonglab/imargi <span class="se">\</span>
    imargi_wrapper.sh <span class="se">\</span>
    -r hg38 <span class="se">\</span>
    -N test_sample <span class="se">\</span>
    -t <span class="m">4</span> <span class="se">\</span>
    -g ./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta <span class="se">\</span>
    -i ./ref/bwa_index/bwa_index_hg38 <span class="se">\</span>
    -1 ./data/sample_R1.fastq.gz <span class="se">\</span>
    -2 ./data/sample_R2.fastq.gz <span class="se">\</span>
    -o ./output
</pre></div>
</div>
<p>We only set a few arguments in the command which are described below, and use default settings for other arguments.</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">-r</span></code>: Name of reference genome. It will be used in header of .pairs format file and names of new generated
reference files.</li>
<li><code class="docutils literal notranslate"><span class="pre">-N</span></code>: Name of sample. It will be used in names of final output files.</li>
<li><code class="docutils literal notranslate"><span class="pre">-t</span></code>: Maximum CPU cores.</li>
<li><code class="docutils literal notranslate"><span class="pre">-g</span></code>: Reference genome sequence FASTA file.</li>
<li><code class="docutils literal notranslate"><span class="pre">-i</span></code>: BWA index files.</li>
<li><code class="docutils literal notranslate"><span class="pre">-1</span></code> and <code class="docutils literal notranslate"><span class="pre">-2</span></code>: Sequencing read pairs FASTQ files.</li>
<li><code class="docutils literal notranslate"><span class="pre">-o</span></code>: Output directory</li>
</ul>
<p>In fact, there are more arguments of <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> can be set, you can check the full arguments in the
<a class="reference internal" href="commandline_api.html"><span class="doc">Command-line API</span></a> section. Here we want to remind you the following three arguments of required
reference files, including chromosome sizes file, bwa index and restriction fragments bed format file. Because we didn’t
set these arguments in the command line, <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> automatically generated them. If you have already got these
files, you can use these arguments and it will save you some time and disk space. When the three arguments are all set
correctly, the <code class="docutils literal notranslate"><span class="pre">-g</span></code> argument can be ignored.</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">-i</span></code>: BWA index files.</li>
<li><code class="docutils literal notranslate"><span class="pre">-c</span></code>: Chromosome sizes file</li>
<li><code class="docutils literal notranslate"><span class="pre">-R</span></code>: Restriction fragments bed format file</li>
</ul>
<p>The following sections are detail illustrations of how <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> works inside of the iMARGI Docker
container.</p>
</div>
<div class="section" id="generating-required-reference-files">
<h2>Generating Required Reference Files<a class="headerlink" href="#generating-required-reference-files" title="Permalink to this headline">¶</a></h2>
<p>As we only input reference genome sequence FASTA file with <code class="docutils literal notranslate"><span class="pre">-g</span></code> argument, so <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> will automatically
generate other required reference files, such as chromosome sizes, bwa index and restriction fragments.</p>
<div class="section" id="generating-chromosome-sizes-file">
<h3>Generating Chromosome Sizes File<a class="headerlink" href="#generating-chromosome-sizes-file" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Description/Tool</th>
<th>Files/Key Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Tool</strong></td>
<td><code>samtools faidx</code></td>
<td><code>&lt;ref_genome_FASTA&gt;</code></td>
</tr>
<tr>
<td><strong>Input</strong></td>
<td>ref genome sequence in FASTA</td>
<td><code>./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta</code></td>
</tr>
<tr>
<td><strong>Output</strong></td>
<td>chromosome sizes file</td>
<td><code>./ref/chrom.sizes.hg38.txt</code></td>
</tr>
</tbody>
</table><p>The <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> uses <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">faidx</span></code> to generate the chromosome sizes file from reference genome sequence FASTA
file. The command lines used in <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/imargi_example/ref
samtools faidx GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta
cat GRCh38_no_alt_analysis_set_GCA_000001405.15.fa.fai <span class="p">|</span>awk <span class="s1">&#39;BEGIN{OFS=&quot;\t&quot;}{print $1,$2}&#39;</span> &gt;chrom.sizes.hg38.txt
</pre></div>
</div>
</div>
<div class="section" id="building-bwa-index">
<h3>Building BWA Index<a class="headerlink" href="#building-bwa-index" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Description/Tool</th>
<th>Files/Key Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Tool</strong></td>
<td><code>bwa index</code></td>
<td><code>-p &lt;idxbase&gt;</code> <br> <code>&lt;ref_genome_FASTA&gt;</code></td>
</tr>
<tr>
<td><strong>Input</strong></td>
<td>ref genome sequence in FASTA</td>
<td><code>./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta</code></td>
</tr>
<tr>
<td><strong>Output</strong></td>
<td>a folder of bwa index files</td>
<td><code>./ref/bwa_index</code></td>
</tr>
</tbody>
</table><p>BWA is used to map sequencing reads to reference genome, so we need to construct bwa index for reference genome first.
Here is the command lines used in <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> ~/imargi_example/refmkdir
bwa index -p ./ref/bwa_index/bwa_index_hg38 ./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta
</pre></div>
</div>
</div>
<div class="section" id="generating-restriction-fragment-file">
<h3>Generating Restriction Fragment File<a class="headerlink" href="#generating-restriction-fragment-file" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Description/Tool</th>
<th>Files/Key Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Tool</strong></td>
<td><code>imargi_rsfrags.sh</code></td>
<td><code>- r &lt;ref_genome_FASTA&gt;</code> <br> <code>-c &lt;chromSize_file&gt;</code> <br> <code>-e &lt;enzyme_name&gt;</code> <br> <code>-C &lt;cut_position&gt;</code></td>
</tr>
<tr>
<td><strong>Input</strong></td>
<td>ref genome sequence in FASTA</td>
<td><code>./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta</code></td>
</tr>
<tr>
<td><strong>Output</strong></td>
<td>restriction fragment BED file</td>
<td><code>./ref/AluI_frags.bed.gz</code></td>
</tr>
</tbody>
</table><p>AluI enzyme is used to digest genome in iMARGI, whose cutting site is AG^CT. The restriction fragments are used for
filtering out read pairs sequenced from non-RNA-DNA-ligation fragments. The <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> uses <code class="docutils literal notranslate"><span class="pre">imargi_rsfrags.sh</span></code>
tool to generate the restriction fragment bed (gzip compressed) format file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>imargi_rsfrags.sh <span class="se">\</span>
    -r ./ref/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta <span class="se">\</span>
    -c ./ref/hg38.chrom.sizes <span class="se">\</span>
    -e AluI <span class="se">\</span>
    -C <span class="m">2</span> <span class="se">\</span>
    -o ./ref/iMARGI_AluI_rsfrags.bed.gz
</pre></div>
</div>
</div>
</div>
<div class="section" id="cleaning-clean-read-pairs">
<h2>Cleaning: clean read pairs<a class="headerlink" href="#cleaning-clean-read-pairs" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Description/Tool</th>
<th>Files/Key Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Tool</strong></td>
<td><code>imargi_clean.sh</code></td>
<td><code>-1 &lt;fastq.gz_R1&gt;</code> <br> <code>-2 &lt;fastq.gz_R2&gt;</code> <br> <code>-o &lt;output_dir&gt;</code> <br> <code>-t &lt;threads&gt;</code></td>
</tr>
<tr>
<td><strong>Input</strong></td>
<td>raw paired FASTQ files</td>
<td><code>./data/sample_R1.fastq.gz</code> <br> <code>./data/sample_R2.fastq.gz</code></td>
</tr>
<tr>
<td><strong>Output</strong></td>
<td>clean paired FASTQ files</td>
<td><code>./output/clean_fastq/clean_test_sample_R1.fastq.gz</code> <br> <code>./output/clean_fastq/clean_test_sample_R2.fastq.gz</code></td>
</tr>
</tbody>
</table><p>According to the design of iMARGI sequencing library construction, there are two random nucleotides <code class="docutils literal notranslate"><span class="pre">NN</span></code> at the 5’ end
of R1 read. It will affect mapping, so we remove those two random bases before mapping.</p>
<p>As the 5’ end of R2 is DNA restriction enzyme digestion site, which is AluI in iMARGI experiment (AG^CT), so the DNA
end should always start with “CT”. Hence, we can use this to filter out non-ligated RNA contaminations.The
<code class="docutils literal notranslate"><span class="pre">imargi_clean.sh</span></code> tool has an option <code class="docutils literal notranslate"><span class="pre">-f</span> <span class="pre">&lt;filter_CT&gt;</span></code>, which can be used to filter read pairs based on the 5’
end sequence of R2 reads as <code class="docutils literal notranslate"><span class="pre">-f</span> <span class="pre">CT</span></code>. The advantage of of such hard filtering is reduce the computational cost.
But the disadvantage is that it’s not flexible and causes false negative. So we don’t use the filter, and the filter
based on restriction fragments is applied in parsing step within the <code class="docutils literal notranslate"><span class="pre">imargi_parse.sh</span></code> tool.</p>
<p>The command of cleaning used in the <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bash imargi_clean.sh <span class="se">\</span>
    -1 ./data/sample_R1.fastq.gz <span class="se">\</span>
    -2 ./data/sample_R2.fastq.gz <span class="se">\</span>
    -o ./output/clean_fastq/ <span class="se">\</span>
    -t <span class="m">16</span>
</pre></div>
</div>
</div>
<div class="section" id="mapping-map-reads-to-genome">
<h2>Mapping: map reads to genome<a class="headerlink" href="#mapping-map-reads-to-genome" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Description/Tool</th>
<th>Files/Key Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Tool</strong></td>
<td><code>bwa mem</code></td>
<td><code>-SP5M</code> <br> <code>-t &lt;threads&gt;</code> <br> <code>&lt;idxbase&gt;</code> <br> <code>&lt;in1.fq&gt;</code> <br> <code>&lt;in2.fq&gt;</code></td>
</tr>
<tr>
<td><strong>Input</strong></td>
<td>clean paired FASTQ files</td>
<td><code>./output/clean_fastq/clean_test_sample_R1.fastq.gz</code> <br> <code>./output/clean_fastq/clean_test_sample_R2.fastq.gz</code></td>
</tr>
<tr>
<td><strong>Output</strong></td>
<td>BAM file</td>
<td><code>./output/bwa_output/test_sample.bam</code></td>
</tr>
</tbody>
</table><p>We use bwa mem with <code class="docutils literal notranslate"><span class="pre">-SP5M</span></code> parameters to map cleaned read pairs to reference genome. The <code class="docutils literal notranslate"><span class="pre">-SP5M</span></code> options are very
important, which allows split alignments. The output BAM file name is based on the <code class="docutils literal notranslate"><span class="pre">-N</span></code> argument of <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code>,
and the name will be kept in next step.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># bwa mem mapping</span>
bwa mem -t <span class="m">16</span> -SP5M ./ref/bwa_index/bwa_index_GRCh38 <span class="se">\</span>
    ./clean_fastq/clean_test_sample_R1.fastq.gz <span class="se">\</span>
    ./clean_fastq/clean_test_sample_R2.fastq.gz <span class="p">|</span><span class="se">\</span>
    samtools view -Shb -@ <span class="m">7</span> - &gt;./bwa_output/test_sample.bam
</pre></div>
</div>
</div>
<div class="section" id="parsing-parse-mapped-read-pairs-to-valid-rna-dna-interactions">
<h2>Parsing: parse mapped read pairs to valid RNA-DNA interactions<a class="headerlink" href="#parsing-parse-mapped-read-pairs-to-valid-rna-dna-interactions" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Description/Tool</th>
<th>Files/Key Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Tool</strong></td>
<td><code>imargi_parse.sh</code></td>
<td><code>-r &lt;ref_name&gt;</code> <br> <code>-c &lt;chromSize_file&gt;</code> <br> <code>-R &lt;restrict_frags_file&gt;</code> <br> <code>-b &lt;bam_file&gt;</code> <br> <code>-o &lt;output_dir&gt;</code> <br> <code>-t &lt;threads&gt;</code></td>
</tr>
<tr>
<td><strong>Input</strong></td>
<td>BAM file</td>
<td><code>.output/bwa_output/test_sample.bam</code></td>
</tr>
<tr>
<td><strong>Output</strong></td>
<td>.pairs file</td>
<td><code>./output/final_test_sample.pairs.gz</code></td>
</tr>
</tbody>
</table><p>The <code class="docutils literal notranslate"><span class="pre">imargi_parse.sh</span></code> is a wrapper script to parse interaction pairs from BAM file, and do de-duplication and filtering,
then output the final .paris file of valid RNA-DNA interaction map, and by default the intermediate files are also kept
in <code class="docutils literal notranslate"><span class="pre">./output/parse_temp</span></code>. There are also some arguments can be used to tweak the thresholds of parsing and filtering.
You can use the default value or customize them.</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">-Q</span> <span class="pre">&lt;min_mapq&gt;</span></code>: The minimal MAPQ score to consider a read as uniquely mapped. [default: 1]</li>
<li><code class="docutils literal notranslate"><span class="pre">-G</span> <span class="pre">&lt;max_inter_align_gap&gt;</span></code>: Read segments that are not covered by any alignment and longer than the specified value
are treated as “null” alignments. [default: 20]</li>
<li><code class="docutils literal notranslate"><span class="pre">-O</span> <span class="pre">offset_restriction_site</span></code>: Offset allowed for restriction fragment filter. [default: 3]</li>
<li><code class="docutils literal notranslate"><span class="pre">-M</span> <span class="pre">&lt;max_ligation_size&gt;</span></code>: Selected sequencing fragment size. [default: 1000]</li>
</ul>
<p>The command of parsing used in the <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>imargi_parse.sh <span class="se">\</span>
    -r hg38 <span class="se">\</span>
    -c ./ref/hg38.chrom.sizes <span class="se">\</span>
    -R ./ref/AluI_frags.bed.gz <span class="se">\</span>
    -b ./output/bwa_output/test_sample.bam <span class="se">\</span>
    -o ./output <span class="se">\</span>
    -t <span class="m">16</span> <span class="se">\</span>
    -Q <span class="m">1</span> <span class="se">\</span>
    -G <span class="m">20</span> <span class="se">\</span>
    -O <span class="m">3</span> <span class="se">\</span>
    -M <span class="m">1000</span> <span class="se">\</span>
    -d <span class="nb">false</span> <span class="se">\</span>
    -D <span class="nv">$output_dir</span>/parse_temp
</pre></div>
</div>
<p>The core tool used in <code class="docutils literal notranslate"><span class="pre">imargi_wrapper.sh</span></code> is <code class="docutils literal notranslate"><span class="pre">pairtools</span></code>. Here are some brief descriptions of the usages of pairtools
in our work. If you want to know more, please check the <a class="reference external" href="https://github.com/mirnylab/pairtools">GitHub repo</a> and
<a class="reference external" href="https://pairtools.readthedocs.io/en/latest/">documentation</a> of <code class="docutils literal notranslate"><span class="pre">pairtools</span></code>.</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">pairtools</span> <span class="pre">parse</span></code>: Parse the types of read pairs based on their mapping results. Output a .pairs file. The
parameters <code class="docutils literal notranslate"><span class="pre">--add-columns</span> <span class="pre">cigar</span></code>, <code class="docutils literal notranslate"><span class="pre">--no-flip</span></code>, <code class="docutils literal notranslate"><span class="pre">--walks-policy</span> <span class="pre">5any</span></code> are required for iMARGI data, don’t change it
unless you know what you want to do.</li>
<li><code class="docutils literal notranslate"><span class="pre">pairtools</span> <span class="pre">dedup</span></code>: Mark and de-duplications in the <code class="docutils literal notranslate"><span class="pre">.pairs</span></code> file.</li>
<li><code class="docutils literal notranslate"><span class="pre">pairtools</span> <span class="pre">select</span></code>: Filter out those read pairs which are marked as duplications, multiple mappings and 5’ most end
was not unique mapped. We used a long filter string designed for iMARGI, don’t change it.</li>
</ul>
<p>There will be <code class="docutils literal notranslate"><span class="pre">pipelineStats_xx.log</span></code> file in the output directory. It reports some basic statistics of the parsing
results, such as total number of read pairs, unique mapped read pairs and valid RNA-DNA interactions.</p>
<blockquote>
<div><p><strong>Filtering strategies</strong></p>
<p>Besides of filtering out unmapped, duplicated and multiple mapped read pairs, we also use
restriction fragments information for filtering out those potential un-ligated RNA or DNA fragment contaminations.
We check the 5’ end of mapped read pairs R1 (RNA-end) and R2 (DNA-end), there are three cases based on the relative
position of 5’ end and the nearest restriction digestion site, which are denoted as dist1 and dist2, and we have a
threshold of offset:</p>
<ul class="simple">
<li>dist2 &gt; offset: R2 (DNA-end) is not sequenced from a proper restriction fragment, so the sequencing fragment might
be un-ligated RNA</li>
<li>dist2 &lt; offset &amp; dist1 &lt; offset: sequencing fragment might be a un-ligated DNA if the distance of R1 and R2 is less
than the maximum sequencing fragment size and the strands of R1 and R2 are opposite</li>
<li>dist2 &lt; offset &amp; dist1 &gt; offset: proper RNA-DNA ligation</li>
</ul>
<p>The filtering argument used in pairtools is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span> <span class="s2">&quot;regex_match(pair_type, \&quot;[UuR][UuR]\&quot;) and \</span>
<span class="s2">    dist2_rsite != \&quot;!\&quot; and \</span>
<span class="s2">    (abs(int(dist2_rsite)) &lt;= &quot;</span><span class="nv">$offset</span><span class="s2">&quot;) and \</span>
<span class="s2">    (not (chrom1 == chrom2 and \</span>
<span class="s2">        abs(int(dist1_rsite)) &lt;= &quot;</span><span class="nv">$offset</span><span class="s2">&quot; and \</span>
<span class="s2">        strand1 != strand2 and \</span>
<span class="s2">        ((strand1 == \&quot;+\&quot; and strand2 == \&quot;-\&quot; and int(frag1_start) &lt;= int(frag2_start) and \</span>
<span class="s2">            abs(int(frag2_end) - int(frag1_start)) &lt;= &quot;</span><span class="nv">$max_ligation_size</span><span class="s2">&quot;) or \</span>
<span class="s2">         (strand1 == \&quot;-\&quot; and strand2 == \&quot;+\&quot; and int(frag1_start) &gt;= int(frag2_start) and \</span>
<span class="s2">            abs(int(frag1_end) - int(frag2_start)) &lt;= &quot;</span><span class="nv">$max_ligation_size</span><span class="s2">&quot;))))&quot;</span>  
</pre></div>
</div>
</div></blockquote>
<p>Although we set several filtering strategies, they still cannot filter out all ‘contaminations’. So we use genomic
distance as a final filter. Filtering out those short-range interactions is better for further analysis. We provide a
tool <code class="docutils literal notranslate"><span class="pre">imargi_distfilter.sh</span></code> to do genomic distance filter. Please read more in the
<a class="reference internal" href="further_analysis.html"><span class="doc">further analysis instructions</span></a>.</p>
</div>
<div class="section" id="output-files">
<h2>Output Files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<div class="section" id="all-the-output-file-list">
<h3>All The Output File List<a class="headerlink" href="#all-the-output-file-list" title="Permalink to this headline">¶</a></h3>
<p>The output files are all in the output directory <code class="docutils literal notranslate"><span class="pre">~/imargi_example/output</span></code>.</p>
<p>The final RNA-DNA interaction map is the <code class="docutils literal notranslate"><span class="pre">final_test_sample.pairs.gz</span></code> file, which is in a compressed <code class="docutils literal notranslate"><span class="pre">.pairs</span></code> format and
can be used for further analysis. There is a pipeline summary log file, <code class="docutils literal notranslate"><span class="pre">pipelineStats_test_sample.log</span></code> in the output
directory, which reports the sequence mapping QC result (pass or failed), total processed read pairs number, BWA mapping
stats and number of valid RNA-DNA interaction in the final <code class="docutils literal notranslate"><span class="pre">.pairs.gz</span></code> file. All other intermediate result files are
also kept in several sub-folders of the output directory.</p>
<p>Besides, if we used the minimum input requirements, the pipeline will automatically generated several required reference files in
the same directory of reference genome instead of the output directory, including chromosome sizes, bwa index and AluI
digestion fragments. When processing new dataset, you can reuse these new generated reference files with corresponding
arguments instead of only using <code class="docutils literal notranslate"><span class="pre">-g</span></code> argument, which will save you some time and disk space.</p>
<p>The final tree structure of the output directory is shown below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    └── output
        ├── bwa_output
        │   ├── bwa_log_test_sample.txt
        │   └── test_sample.bam
        ├── clean_fastq
        │   ├── clean_test_sample_R1.fastq.gz
        │   └── clean_test_sample_R2.fastq.gz
        ├── parse_temp
        │   ├── dedup_test_sample.pairs.gz
        │   ├── drop_test_sample.pairs.gz
        │   ├── duplication_test_sample.pairs.gz
        │   ├── sorted_all_test_sample.pairs.gz
        │   ├── stats_dedup_test_sample.txt
        │   ├── stats_final_test_sample.txt
        │   └── unmapped_test_sample.pairs.gz
        ├── final_test_sample.pairs.gz
        └── pipelineStats_test_sample.log
</pre></div>
</div>
<p>The table below briefly described all the output files.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Type</th>
<th>Example Items</th>
<th>Directory (relative to output directory)</th>
<th>Format</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cleaned FASTQ</td>
<td><code>clean_test_sample_R1.fastq.gz</code> <br> <code>clean_test_sample_R2.fastq.gz</code></td>
<td><code>./clean_fastq</code></td>
<td>paired clean FASTQ</td>
<td>Output of cleaning step, the two random bases at 5' end of R1 were removed</td>
</tr>
<tr>
<td>Sequencing reads alignments</td>
<td><code>test_sample.bam</code></td>
<td><code>./bwa_output</code></td>
<td>BAM</td>
<td>Output of bwa mapping step</td>
</tr>
<tr>
<td>BWA mapping log</td>
<td><code>bwa_log_test_sample.txt</code></td>
<td><code>./bwa_output</code></td>
<td>text</td>
<td>BWA mapping log file</td>
</tr>
<tr>
<td>Intermediate parsing results</td>
<td><code>sorted_all_test_sample.pairs.gz</code></td>
<td><code>./parse_temp</code></td>
<td>.pairs</td>
<td>Sorted all the parsed read pairs</td>
</tr>
<tr>
<td>Intermediate parsing results</td>
<td><code>dedup_test_sample.pairs.gz</code></td>
<td><code>./parse_temp</code></td>
<td>.pairs</td>
<td>Parsed read pairs after deduplication</td>
</tr>
<tr>
<td>Intermediate parsing results</td>
<td><code>duplication_test_sample.pairs.gz</code></td>
<td><code>./parse_temp</code></td>
<td>.pairs</td>
<td>Duplicated pairs</td>
</tr>
<tr>
<td>Intermediate parsing results</td>
<td><code>unmapped_test_sample.pairs.gz</code></td>
<td><code>./parse_temp</code></td>
<td>.pairs</td>
<td>Unmapped pairs (including multiple mapping)</td>
</tr>
<tr>
<td>Intermediate parsing results</td>
<td><code>drop_test_sample.pairs.gz</code></td>
<td><code>./parse_temp</code></td>
<td>.pairs</td>
<td>Pairs filtered out filtering step</td>
</tr>
<tr>
<td>Intermediate parsing results</td>
<td><code>stats_dedup_test_sample.txt</code> <br>  <code>stats_final_test_sample.txt</code></td>
<td><code>./parse_temp</code></td>
<td>text</td>
<td>pairtools intermediate stats report</td>
</tr>
<tr>
<td>Final output</td>
<td><code>final_test_sample.pairs.gz</code></td>
<td><code>./</code></td>
<td>.pairs</td>
<td>Final output of valid RNA-DNA interactions</td>
</tr>
<tr>
<td>Pipeline stats log</td>
<td><code>pipelineStats_test_sample.log</code></td>
<td><code>./</code></td>
<td>text</td>
<td>Sequence mapping QC result and stats report</td>
</tr>
</tbody>
</table></div>
<div class="section" id="the-imargi-pipeline-output-pairs-format">
<h3>The iMARGI Pipeline Output <code class="docutils literal notranslate"><span class="pre">.pairs</span></code> Format<a class="headerlink" href="#the-imargi-pipeline-output-pairs-format" title="Permalink to this headline">¶</a></h3>
<p>The default final output file is <code class="docutils literal notranslate"><span class="pre">final_test_sample.pairs.gz</span></code> which is in .pairs format. .pairs file format is designed
by 4DN DCIC. You can read its
<a class="reference external" href="https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification">specification document</a>.</p>
<p>In iMARGI Pipeline, we add some more data columns to the default .pairs format files.</p>
<p><img alt="_images/pairs_cols.png" src="_images/pairs_cols.png" /></p>
<p>The data column descriptions are listed in the table below. (All the genomic coordinates in .pairs file are 1-based.)
The first 7 columns are standard information, 8-18 are extra information. All these column names are reserved.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>column order</th>
<th>name</th>
<th>datatype</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>readID</td>
<td>string</td>
<td>read ID</td>
</tr>
<tr>
<td>2</td>
<td>chrom1</td>
<td>string</td>
<td>mapping chromosome of R1 (RNA-end)</td>
</tr>
<tr>
<td>3</td>
<td>pos1</td>
<td>int</td>
<td>5' end mapping position of R1 (RNA-end)</td>
</tr>
<tr>
<td>4</td>
<td>chrom2</td>
<td>string</td>
<td>mapping chromosome of R2 (DNA-end)</td>
</tr>
<tr>
<td>5</td>
<td>pos2</td>
<td>int</td>
<td>5' end mapping position of R2 (DNA-end)</td>
</tr>
<tr>
<td>6</td>
<td>strand1</td>
<td>+/-</td>
<td>mapping strand of R1 (RNA-end), the actual cDNA strand is opposite</td>
</tr>
<tr>
<td>7</td>
<td>strand2</td>
<td>+/-</td>
<td>mapping strand of R2 (DNA-end)</td>
</tr>
<tr>
<td>8</td>
<td>pair_type</td>
<td>string</td>
<td>pairtools defined pair_type</td>
</tr>
<tr>
<td>9</td>
<td>mapq1</td>
<td>int</td>
<td>bwa MAPQ value of R1 (RNA-end) alignment</td>
</tr>
<tr>
<td>10</td>
<td>mapq2</td>
<td>int</td>
<td>bwa MAPQ value of R2 (DNA-end) alignment</td>
</tr>
<tr>
<td>11</td>
<td>cigar1</td>
<td>string</td>
<td>bwa cigar of R1 (RNA-end) alignment</td>
</tr>
<tr>
<td>12</td>
<td>cigar2</td>
<td>string</td>
<td>bwa cigar of R2 (DNA-end) alignment</td>
</tr>
<tr>
<td>13</td>
<td>frag1_start</td>
<td>int</td>
<td>start position of assigned restriction fragment of R1 (RNA-end)</td>
</tr>
<tr>
<td>14</td>
<td>frag1_end</td>
<td>int</td>
<td>end position of assigned restriction fragment of R1 (RNA-end)</td>
</tr>
<tr>
<td>15</td>
<td>dist1_rsite</td>
<td>int</td>
<td>distance between 5' end mapping position of R1 and the nearest restriction digestion site position</td>
</tr>
<tr>
<td>16</td>
<td>frag2_start</td>
<td>int</td>
<td>start position of assigned restriction fragment of R2 (DNA-end)</td>
</tr>
<tr>
<td>17</td>
<td>frag2_end</td>
<td>int</td>
<td>end position of assigned restriction fragment of R2 (DNA-end)</td>
</tr>
<tr>
<td>18</td>
<td>dist2_rsite</td>
<td>int</td>
<td>distance between 5' end mapping position of R2 and the nearest restriction digestion site position</td>
</tr>
</tbody>
</table></div>
<div class="section" id="the-pipeline-summary-log-file">
<h3>The pipeline summary log file<a class="headerlink" href="#the-pipeline-summary-log-file" title="Permalink to this headline">¶</a></h3>
<p>When the pipeline finished, it will also generate a simple pipeline summary log file, named as <code class="docutils literal notranslate"><span class="pre">pipelineStats_*.log</span></code>. It
reports the sequencing mapping QC result at the first line. If it shows “Sequence mapping QC failed”, your data might
have critical problem for further analysis. The other lines show stats number of total read pairs,
non-dup unique mapping read pairs, final valid read pairs and so on.</p>
<p>Here we describe each line of the log file (TAB separated text file).</p>
<ul>
<li><p class="first">Line 1: <code class="docutils literal notranslate"><span class="pre">Sequence</span> <span class="pre">mapping</span> <span class="pre">QC</span></code></p>
<p>It’s the sequencing mapping QC result, which can be <code class="docutils literal notranslate"><span class="pre">pass</span></code> or <code class="docutils literal notranslate"><span class="pre">failed</span></code>). It’s concluded based on the value of line 2
and 3. Only when both values of line 2 and 3 are greater than threshold 0.5, it can be concluded as <code class="docutils literal notranslate"><span class="pre">pass</span></code>.</p>
</li>
<li><p class="first">Line 2: <code class="docutils literal notranslate"><span class="pre">(#unique_mapped_pairs</span> <span class="pre">+</span> <span class="pre">#single_side_unique_mapped)/#total_read_pairs</span></code></p>
<p>Mapping rate of at least one side of read pairs uniquely mapped to genome. It can be calculated by dividing the sum of
line 5 and line 6 by line 4.</p>
</li>
<li><p class="first">Line 3: <code class="docutils literal notranslate"><span class="pre">#total_valid_interactions/#unique_mapped_pairs</span></code></p>
<p>The ratio of retained read pairs of unique mapped read pairs after de-duplication and filtering. It can be calculated
by dividing line 8 by line 6.</p>
</li>
<li><p class="first">Line 4: <code class="docutils literal notranslate"><span class="pre">total_read_pairs</span></code></p>
<p>Total number of read pairs in the FASTQ data.</p>
</li>
<li><p class="first">Line 5: <code class="docutils literal notranslate"><span class="pre">single_side_unique_mapped</span></code></p>
<p>Number of single side unique mapped read pairs.</p>
</li>
<li><p class="first">Line 6: <code class="docutils literal notranslate"><span class="pre">unique_mapped_pairs</span></code></p>
<p>Number of unique mapped read pairs (both two side unique mapped).</p>
</li>
<li><p class="first">Line 7: <code class="docutils literal notranslate"><span class="pre">non_dup_unique_mapped_paris</span></code></p>
<p>Number of non-duplicated unique mapped read pairs.</p>
</li>
<li><p class="first">Line 8: <code class="docutils literal notranslate"><span class="pre">total_valid_interactions</span></code></p>
<p>Number of final output valid read pairs (RNA-DNA interactions).</p>
</li>
<li><p class="first">Line 9: <code class="docutils literal notranslate"><span class="pre">inter_chr</span></code></p>
<p>Number of inter-chromosomal interacted read pairs (two ends mapped to different chromosomes).</p>
</li>
<li><p class="first">Line 10: <code class="docutils literal notranslate"><span class="pre">intra_chr</span></code></p>
<p>Number of intra-chromosomal interacted read pairs (two ends mapped to the same chromosome).</p>
</li>
</ul>
<p>Here is the output <code class="docutils literal notranslate"><span class="pre">pipelineStats_test_sample.log</span></code> file in our example.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Sequence</span> <span class="n">mapping</span> <span class="n">QC</span> <span class="n">failed</span>
<span class="p">(</span><span class="c1">#unique_mapped_pairs + #single_side_unique_mapped)/#total_read_pairs   0.777777</span>
<span class="c1">#total_valid_interactions/#unique_mapped_pairs  0.761476</span>
<span class="n">total_read_pairs</span>    <span class="mi">900000</span>
<span class="n">single_side_unique_mapped</span>   <span class="mi">2</span>
<span class="n">unique_mapped_pairs</span> <span class="mi">699997</span>
<span class="n">non_dup_unique_mapped_paris</span> <span class="mi">691967</span>
<span class="n">total_valid_interactions</span>    <span class="mi">533031</span>
<span class="n">inter_chr</span>   <span class="mi">244163</span>
<span class="n">intra_chr</span>   <span class="mi">288868</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="further_analysis.html" class="btn btn-neutral float-right" title="Further Analysis and Visualization Guides" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="quick_example.html" class="btn btn-neutral" title="Quick Start Example" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Zhong Lab, UCSD. iMARGI-Docker code licensed under the BSD 2 License. Documentation licensed under CC BY 3.0

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'1.1',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>